Add SetRuntimeEnvironment unit tests

ni-jfitzger opened the pull request:
### What does this Pull Request accomplish?

* Add unit tests for SetRuntimeEnvironment behavior.
* Add a SharedLibraryInterface class and MockSharedLibrary
  * Needed to support the unit tests
  * Was not able to figure out a way to include the swap method, but it doesn't seem to be used anywhere

Impelmentation Notes:
* Tests were implemented with NI-SWITCH, instead of NI-FAKE due to functions missing from nifake.h that were present in the metadata.
  * I looked into adding these to nifake.h, but after iterating on the PR several times (having to add more and more: attributes, enums, etc, I decided it wasn't worth the effort).
* Updated Library Classes
  * They now take in a shared_ptr<SharedLibraryInterface> instead of creating their own SharedLibrary object, to support the unit tests
 * Another constructor was added to automatically create the shared_pointer<SharedLibrary> object to preserve existing behavior
  * Added an attribute and method to track, determine whether SetRuntimeEnvironment has been called (for testing purposes)


### Why should this Pull Request be merged?

There are no plans to add system-level testing for SetRuntimeEnvironment in grpc-device, so this is the only testing we'll have to make sure it doesn't get broken.

### What testing has been done?

Ran unit tests locally

bkeryan commented on source/codegen/templates/library.cpp.mako:
```diff
@@ -15,18 +15,26 @@ linux_so_version_suffix = f'.{linux_abi_version}' if linux_abi_version else ''
 
+class_name = f"{service_class_prefix}Library"
+
+intialization_list = "p_shared_library_(pSharedLibrary)"
```
spelling

bkeryan commented on source/codegen/templates/library.cpp.mako:
```diff
@@ -35,10 +43,13 @@ static const char* kLibraryName = "lib${linux_library_name}.so${linux_so_version
+
+${class_name}::${class_name}(std::shared_ptr<nidevice_grpc::SharedLibraryInterface> pSharedLibrary) : ${intialization_list}
 {
-  shared_library_.load();
-  bool loaded = shared_library_.is_loaded();
+  p_shared_library_->set_library_name(kLibraryName);
```
Why 'p'?

ni-jfitzger replied:
Because it's a pointer.
I'm open to changing the name.

bkeryan replied:
Does any other pointer variable in this codebase have a "p" prefix? When in Rome, do as the Romans do.

ni-jfitzger replied:
Updated the name for the attribute and the constructor parameter.

bkeryan commented on source/codegen/templates/library.h.mako:
```diff
@@ -60,8 +70,11 @@ class ${service_class_prefix}Library : public ${namespace_prefix}::${service_cla
+  std::shared_ptr<nidevice_grpc::SharedLibraryInterface> p_shared_library_;
   FunctionPointers function_pointers_;
+% if set_runtime_environment_supported:
+  bool runtime_environment_set; // needed to test that we properly call SetRuntimeEnvironment
```
Missing trailing underscore

bkeryan commented on source/codegen/templates/library.h.mako:
```diff
@@ -37,6 +44,9 @@ class ${service_class_prefix}Library : public ${namespace_prefix}::${service_cla
   ${return_type} ${method_name}(${service_helpers.create_params(parameters, expand_varargs=True)});
 % endfor
+% if set_runtime_environment_supported:
+  bool get_runtime_environment_set(); // needed to test that we properly call SetRuntimeEnvironment
```
"is"/"was" sounds more natural for Booleans than "get".

bkeryan commented on source/tests/unit/library_set_runtime_environment_tests.cpp:
```diff
@@ -0,0 +1,63 @@
+{
+  auto p_shared_library = std::make_shared<MockSharedLibrary>();
+  EXPECT_CALL(*p_shared_library, set_library_name(_))
+    .Times(1);
```
`.Times(1)` is the default for most of these calls.
https://google.github.io/googletest/gmock_for_dummies.html#cardinalities-how-many-times-will-it-be-called

bkeryan commented on source/tests/unit/library_set_runtime_environment_tests.cpp:
```diff
@@ -0,0 +1,63 @@
+  ON_CALL(*p_shared_library, get_function_pointer(func_name))
+    .WillByDefault(Return(&FunctionToPointAt));
+  EXPECT_CALL(*p_shared_library, get_function_pointer(_))
+    .Times(AtLeast(1));
```
You can simplify this by removing the `ON_CALL` and using `.WillRepeatedly(Return(...))` or `.WillOnce(Return(...))` on the `EXPECT_CALL`.
https://google.github.io/googletest/gmock_for_dummies.html#cardinalities-how-many-times-will-it-be-called

`ON_CALL` is more commonly used to specify default behavior in a setup/constructor or to specify behavior for `NiceMock`.

ni-jfitzger replied:
I'll give it a shot. I actually tried that first and couldn't seem to get it working.

reckenro reviewed (approved):
Looks good to me. And Alex is out through next week so no need to wait on his review.

ni-jfitzger commented:
> Linux CI failures:
> 
> * `/home/runner/work/grpc-device/grpc-device/./generated/niswitch/niswitch_library.h:20:3: error: extra qualification ‘niswitch_grpc::NiSwitchLibrary::’ on member ‘NiSwitchLibrary’ [-fpermissive]`
> * `/home/runner/work/grpc-device/grpc-device/third_party/gtest/googlemock/include/gmock/gmock-actions.h:682:40: error: invalid conversion from ‘void (*)()’ to ‘const void*’ [-fpermissive]`

Removed the unnecessary qualifer and added const_cast<const void*>(reinterpret_cast<void*>())

bkeryan commented on generated/nidaqmx/nidaqmx_library.h:
```diff
@@ -17,7 +17,7 @@ namespace nidaqmx_grpc {
  public:
   NiDAQmxLibrary();
-  NiDAQmxLibrary::NiDAQmxLibrary(std::shared_ptr<nidevice_grpc::SharedLibraryInterface>);
+  NiDAQmxLibrary(std::shared_ptr<nidevice_grpc::SharedLibraryInterface>);
```
Two minor nitpicks I didn't notice before:
- The parameter is unnamed
- This single-parameter constructor is missing `explicit`, so it enables implicit conversion from `shared_ptr<SharedLibraryInterface>` to `NiDAQmxLibrary&`, etc.

ni-jfitzger replied:
>The parameter is unnamed

Is it not common to just specify parameter types in the function declaration?
I can add the name.

>This single-parameter constructor is missing explicit, so it enables implicit conversion from shared_ptr<SharedLibraryInterface> to NiDAQmxLibrary&, etc.

Nice catch.

bkeryan replied:
> Is it not common to just specify parameter types in the function declaration?

I usually see unnamed parameters used to work around unused variable warnings.

ni-jfitzger commented:
@reckenro @bkeryan Are the systems tests waiting for approval from an owner before running?

bkeryan commented:
> @reckenro @bkeryan Are the systems tests waiting for approval from an owner before running?

No, this is an intermittent issue with the self-hosted GitHub runner. I'll send you more info offline.
